### PR TITLE
Fix reverse DNAT for externalTrafficPolicy=Local service with kube-proxy

### PR DESCRIPTION
**Pull request by pchaigno**
See commits for details.

I'll rebase on https://github.com/cilium/cilium/pull/14675 once merged (the two PRs share the same first commit).

EDIT: The release note describes the fix for the host firewall. The regression for IPSec didn't make it into a release, so no need for a release note.
```release-note
Fix connectivity to externalTrafficPolicy=Local services when using the host firewall with kube-proxy
```

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-4.9

**Comment by pchaigno** [comment]
test-netnext

**Comment by pchaigno** [comment]
test-runtime

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-gke

**Comment by pchaigno** [comment]
test-me-please

File: pkg/datapath/iptables/iptables.go [hunk 1]
```diff
@@ -1016,7 +1016,7 @@ func (m *IptablesManager) installMasqueradeRules(prog, ifName, localDeliveryInte
 			m.waitArgs,
 			"-t", "nat",
 			"-A", ciliumPostNatChain,
-			"!", "-s", hostMasqueradeIP,
+			"!", "-s", allocRange,
```
Comment (joestringer) [inline pkg/datapath/iptables/iptables.go:1016]: Had to think about it for a bit, but LGTM. The comment above needs updating though I think, "if the source is not the internal IP" is the part that's changing here. It also mentions tunnel and non-tunnel modes, but this rule installation logic is only handled in the tunneling case.

**Review by joestringer — APPROVED** [review]
Minor nits to address.

**Review by aditighag — APPROVED** [review]

**Comment by pchaigno** [comment]
test-gke

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-gke

**Comment by joestringer** [comment]
GKE hit known flake #14915 . With CI signoff, this should be good to merge. @christarazi  / @michi-covalent could you check over the CI side here?

**Review by michi-covalent — APPROVED** [review]

**Comment by pchaigno** [comment]
I updated the release note given this PR also fixed https://github.com/cilium/cilium/issues/12542.

**Pull request merged**